### PR TITLE
Added general readme to type inference

### PR DESCRIPTION
**Pull request by yuanzexiong**
## Description

Added general readme to type inference.

## Changes

1. New README in type inference

## Deletes Source Branch?

Yes:    Work is done, the branch can be safely deleted.

## How Many Approvals?

* 1 Approval (Small bugs/hot-fixes)

### CODE REVIEW

File: src/typeInference/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+The types of each node in the AST need to be determined so that they can later be used
+ for optimisation purposes.
+
+Annotation starts by enabling the `annotateAST` flags in the options parameter of the
+ `evaluateXPath` functions. When the flag is enabled, the `annotateAST` function gets
+ triggered and it will recursively perform a depth-first, pre-order traversal to
+ annotate sub-tree of the AST. Given the type of the sub-tree, the corresponding
+ annotation function gets called. These corresponding functions perform the annotation
+ by first determine the type of the node and then inserting a `SequenceType` to the
```
Comment (FBroy) [inline src/typeInference/README.md]: determining*

File: src/typeInference/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ AST as an attribute called `type`. 
+
+There are cases where the variable references and function references are being used.
+
+- The functional references are being resolved by first extracting the `QName` out of
+ the AST and then perform a lookup in the `StaticContext` (passed as an optional
```
Comment (FBroy) [inline src/typeInference/README.md]: performing*

File: src/typeInference/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ attribute of the `AnnotationContext` class) to determine the arity, function name and
+ the return type of the function.
+
+- The variable references makes use of the `AnnotationContext` that tracks different
+ level of scopes and manages variables in each of the scopes. It is mostly used in the
```
Comment (FBroy) [inline src/typeInference/README.md]: levels*

**Review by HectorPeeters — APPROVED** [review]
Thats a nice start of the documentation!

File: src/typeInference/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+Using the type information introduced by the annotation process, the type of each node
+ can be used to increase efficiency during the evaluation process.
+
+### Early Returns
+
+Once the types are known, certain checks can be skipped to increase the speed.
+
+For instance: binary operators and unary operators can benefit from skipping the type
+ checking once we know what is the types of the operands and operators. If the type
+ information is insufficient, the type is the previous code of determining types is
```
Comment (FBroy) [inline src/typeInference/README.md]: the type of* the previous code

**Review by FBroy — CHANGES_REQUESTED** [review]
Next to the few typos, good job!

**Review by FBroy — APPROVED** [review]

File: src/typeInference/README.md (hunk 2 above)
Comment (yuanzexiong) [reply src/typeInference/README.md]: Fixed

File: src/typeInference/README.md (hunk 1 above)
Comment (yuanzexiong) [reply src/typeInference/README.md]: Fixed

File: src/typeInference/README.md (hunk 3 above)
Comment (yuanzexiong) [reply src/typeInference/README.md]: Fixed

File: src/typeInference/README.md (hunk 4 above)
Comment (yuanzexiong) [reply src/typeInference/README.md]: Paraphrased

**Pull request merged**